[MRG] DOC: OS dependent installation instructions

cmarmo opened the pull request:
#### Reference Issues/PRs
Closes #5737

#### What does this implement/fix? Explain your changes.
Implement interactive installation instructions as in
https://spacy.io/usage/#quickstart

#### Any other comments?
No javascript, only CSS3.
Some selector combination still unmanaged, default styling applied

cmarmo committed:
Start work on html install.

cmarmo committed:
Merge branch 'master' into osdepinstall

cmarmo committed:
Working on os dependent install instructions.

rth commented:
This is quite nice! Maybe the button et checkboxes need a bit more CSS style as right now they look a bit austere )

cmarmo commented:
> This is quite nice! Maybe the button et checkboxes need a bit more CSS style as right now they look a bit austere )

Thanks @rth ... but yes css is still WIP (this time I mean it... :) )

ogrisel commented:
Thanks @cmarmo for bearing with the decision process. Here is a screenshot of the current state:

![Screenshot from 2019-10-09 11-56-32](https://user-images.githubusercontent.com/89061/66471588-dff15680-ea8b-11e9-891f-e0dfa2037fee.png)

I am still +1 for merge :)

rth commented:
@ogrisel Wrong issue :)

ogrisel commented:
oops, sorry wrong tab indeed. I moved my comment to the right PR.

ogrisel commented on doc/install.rst:
```diff
@@ -28,6 +28,51 @@ There are different ways to install scikit-learn:
+          <span class="sk-expandable linux-source">export PYTHONPATH=`pwd`</span>
+          <span class="sk-expandable win-source">set PYTHONPATH=/path/to/scikit-learn</span>
+          <span class="sk-expandable source">pip install -r requirements.txt</span>
+          <span class="sk-expandable source">python install --editable .</span>
```
```suggestion
          <span class="sk-expandable source">pip install --editable .</span>
```

cmarmo replied:
Well... I have finally removed the "source" section, because we agreed that this is the "last release" documentation ... and because it made my life easier... :)

ogrisel replied:
Yes. We can always decide to re-add later or add a similar "quickstart" tool dedicated to building from source to the advanced installation page.

ogrisel commented on doc/install.rst:
```diff
@@ -28,6 +28,51 @@ There are different ways to install scikit-learn:
+          <label for="quickstart-source">source</label>
+       <p><strong>Pip Virtual Environment</strong></p>
+          <input type="checkbox" name="config" id="quickstart-venv">
+          <label for="quickstart-venv">Virtual environment</label>
```
There could be a single checkbox "Use virtualenv / conda environment"  display the `python -m venv sklearn-venv` or  `conda create -n sklearn-env` command depending on the value of the `packager` radio button.

ogrisel replied:
Maybe change the OS order by popularity: "Windows", "macOS", "Linux (and other POSIX systems)"

cmarmo replied:
Done in b8f44c4023a0c5eb1d8a6107defed15bcf4d1166

ogrisel commented on doc/install.rst:
```diff
@@ -28,6 +28,51 @@ There are different ways to install scikit-learn:
+       </code>
+       </pre></div>
+  </div>
+
```
I would add:

Check your installation:

[when "packager" is "pip" or "source"]
```
python -m pip show scikit-learn  # to see which version and where scikit-learn is installed
python -m pip freeze  # to see all packages installed in the active virtualenv
```

[when "package" is "conda"]
```
conda list scikit-learn  # to see which scikit-learn version is installed
conda list  # to see all packages installed in the active conda environment
```

Using an isolated environment (virtualenv or conda environment) makes it possible to install a specific version of scikit-learn and its dependencies independently of any previously installed Python packages. In particular under Linux is it discouraged to install pip packages alongside the packages managed by the package manager of the distribution (apt, dnf, pacman...).

Note that you should always remember to activate the environment of your choice prior to running any Python command whenever you start a new terminal session.

cmarmo replied:
Addressed in the latest commits.

ogrisel commented on doc/install.rst:
```diff
@@ -28,6 +28,51 @@ There are different ways to install scikit-learn:
+
+Read more
+---------
+
```
Instead of a `Read more` section I think we should remove the redundant instructions.

cmarmo replied:
addressed in e373104f3abcb34291e908a4313e79e7b787243f

ogrisel commented on doc/install.rst:
```diff
@@ -28,6 +28,51 @@ There are different ways to install scikit-learn:
+          <span class="sk-expandable source">git clone https://github.com/scikit-learn/scikit-learn</span>
+          <span class="sk-expandable source">cd scikit-learn</span>
+          <span class="sk-expandable linux-source">export PYTHONPATH=`pwd`</span>
+          <span class="sk-expandable win-source">set PYTHONPATH=/path/to/scikit-learn</span>
```
There is no need to set `PYTHONPATH` when building from source.

cmarmo replied:
See above... "source" section has been removed

ogrisel reviewed (commented):
Great PR! Some early feedback:

ogrisel commented on doc/install.rst:
```diff
@@ -28,6 +28,51 @@ There are different ways to install scikit-learn:
+  <div class="install">
+       <p><strong>System</strong></p>
+          <input type="radio" name="os" id="quickstart-mac" checked>
+          <label for="quickstart-mac">macOS / OSX</label>
```
I would remove the "/ OSX" part of the label. macOS is enough (and the OSX name does not exist anymore :). Let's keep things simple to improve readability.

cmarmo replied:
Done in b8f44c4023a0c5eb1d8a6107defed15bcf4d1166

cmarmo committed:
Merge branch 'master' into osdepinstall

cmarmo committed:
Merge branch 'master' into osdepinstall

cmarmo committed:
Merge branch 'master' into osdepinstall

cmarmo committed:
Command switch working. CSS still WIP.

cmarmo commented:
...next time I will use javascript... ;)

cmarmo committed:
Styling buttons. Remove section headers.

cmarmo committed:
Add installation checks.

cmarmo committed:
Merge master.

cmarmo committed:
Finalize installation checks, fix code padding.

ogrisel commented on doc/install.rst:
```diff
@@ -28,24 +28,85 @@ There are different ways to install scikit-learn:
+       <span class="sk-expandable" data-packager="pip" data-os="mac" data-venv="no">python -m pip show scikit-learn  # to see which version and where scikit-learn is installed</span>
+       <span class="sk-expandable" data-packager="pip" data-os="mac" data-venv="no">python -m pip freeze  # to see all packages installed in the active virtualenv</span>
+       <span class="sk-expandable" data-packager="conda">conda list scikit-learn  # to see which scikit-learn version is installed</span>
+       <span class="sk-expandable" data-packager="conda">conda list  # to see all packages installed in the active conda environment</span>
```
One could also add the following command:

```
python -c "import sklearn; sklearn.show_versions()"
```

that would become the following under Linux without venv/conda env:

```
python3 -c "import sklearn; sklearn.show_versions()"
```

cmarmo replied:
Done in 16eff3a94c7c011211bb774ffdd854e2e568eeac

ogrisel replied:
Those instructions should stay displayed when the conda package manager is selected.

ogrisel reviewed (commented):
Nice boolean CSS skills :)

Here is some feedback:

ogrisel commented on doc/install.rst:
```diff
@@ -28,24 +28,85 @@ There are different ways to install scikit-learn:
+       <strong>Use virtualenv / conda environment</strong>
+          <input type="checkbox" name="config" id="quickstart-venv">
+          <label for="quickstart-venv">Yes / No</label>
+       </span>
```
Then one could remove the redundant information from below.

cmarmo replied:
Done in 16eff3a94c7c011211bb774ffdd854e2e568eeac

ogrisel commented on doc/themes/scikit-learn-modern/static/css/theme.css:
```diff
@@ -1049,6 +1048,72 @@ div.testimonial span.testimonial-author p {
+  display: none;
+  background: #ddd;
+}
+
```
Since this expandable / quickstart CSS rules are heavily inspired by the spaCy quickstart page, I think we should:

- ask spaCy authors (ping @ines @honnibal) if they are fine with us being "very inspired" by their work :)
- if so add some credits as a comment in our CSS file.

ogrisel replied:
Either CSS or HTML or both.

cmarmo replied:
I've added a line in both, waiting for feedback.

honnibal replied:
We got the idea for the installation widget from PyTorch initially, so definitely glad to have it spreading! If you haven't seen it, Ines has a repo for the component: https://github.com/ines/quickstart . You don't necessarily have to use the component from the repo of course, but it's probably the right thing to reference in the comments.

cmarmo replied:
Thanks for your answer, @honnibal. I've added the ref to the original Ines implementation in the code.

ogrisel reviewed (commented):
Before merging this PR:

cmarmo committed:
Fix code rendering.

cmarmo committed:
Address comments.

cmarmo committed:
Add credits to spacy.io

cmarmo commented on doc/install.rst:
```diff
@@ -28,24 +28,85 @@ There are different ways to install scikit-learn:
+          <input type="checkbox" name="config" id="quickstart-venv">
+          <label for="quickstart-venv">Yes / No</label>
+       </span>
+       <div><pre>
```
Fixed in ed4521390382907f385c889e2518bf99201be814

ogrisel commented:
It would be nice to find a way to have a `$` shell prompt for all the terminal commands for this page in CSS so that it cannot be selected by copy and and paste.

Currently some terminal command have the prompt, other do not. And those who have, have it selectable.

ogrisel commented:
Later in this page: the URL of pkgsrc-wip should be updated to http://pkgsrc.org/wip/ and the URL o the scikit-learn package is now http://pkgsrc.se/math/py-scikit-learn.

thomasjpfan reviewed (commented):
For buttons, the cursor should change on hover. (By default it is a hand with a glove on it).

cmarmo committed:
Merge branch 'master' into osdepinstall

ogrisel commented on doc/install.rst:
```diff
@@ -28,24 +28,96 @@ There are different ways to install scikit-learn:
+      ><span class="sk-expandable" data-packager="pip" data-os="linux" data-venv="no">python3 -c "import sklearn; sklearn.show_versions()"</span
+      ><span class="sk-expandable" data-packager="pip" data-venv="">python -m pip show scikit-learn  # to see which version and where scikit-learn is installed</span
+      ><span class="sk-expandable" data-packager="pip" data-venv="">python -m pip freeze  # to see all packages installed in the active virtualenv</span
+      ><span class="sk-expandable" data-packager="pip" data-venv="">python -c "import sklearn; sklearn.show_versions()"</span
```
This `python -c "import sklearn; sklearn.show_versions()"` should also be visible when conda is the package manager (with or without conda env).

cmarmo replied:
> For buttons, the cursor should change on hover. (By default it is a hand with a glove on it).

Fixed in f0be47c995696dae075965ad1b7240b2793b3e7a

cmarmo replied:
Done in f0be47c995696dae075965ad1b7240b2793b3e7a, but not sure if on linux with conda python 3 is `python` or `python3`.

ogrisel replied:
When using conda on Linux, it actually depends on whether or not you decide to activate the `base` environment in `.bashrc`. One can assume yes to keep things simple.

cmarmo committed:
Address comments.

cmarmo commented:
> For buttons, the cursor should change on hover. (By default it is a hand with a glove on it).

Fixed in f0be47c995696dae075965ad1b7240b2793b3e7a

cmarmo commented:
> It would be nice to find a way to have a `$` shell prompt for all the terminal commands for this page in CSS so that it cannot be selected by copy and and paste.
> 
> Currently some terminal command have the prompt, other do not. And those who have, have it selectable.

I've added the prompt via css in the quickstart section. I'm not comfortable in looking for general solutions for other command lines because there are probably command lines in other .rst and I don't want to generate unexpected behaviours. So I have just added the prompt as text... for now.

cmarmo committed:
Fix NetBSD skl repository.

cmarmo commented:
> Later in this page: the URL of pkgsrc-wip should be updated to http://pkgsrc.org/wip/ and the URL o the scikit-learn package is now http://pkgsrc.se/math/py-scikit-learn.

Fixed in a850b042dad000592bbd6649a4e8023e49e5d0aa

thomasjpfan commented:
Is it possible to change the text `Use virtualenv / conda environment` depending on the selected Packager?

cmarmo committed:
Merge branch 'master' into osdepinstall

cmarmo committed:
Add reference to original javascript implementation. Switch label between conda and pip.

cmarmo commented:
> Is it possible to change the text `Use virtualenv / conda environment` depending on the selected Packager?

Yes. :)

ogrisel reviewed (approved):
LGTM, thank you very much for this nice onboarding / usability improvement.
